lib/SlackBot.js

```diff
@@ -0,0 +1,331 @@
+  this._token = options.token;
+  this._configUrl = options.configUrl;
+  this._repositories = options.repositories;
+  this._isDisabled = false;
+  this._userIDs = {};
+  this._userData = {};
+  this._channelIDs = {};
+  if (!defined(this._token)) {
+    dateLog('No Slack token found. SlackBot is disabled.');
+    this._isDisabled = true;
+    return;
+  }
+
+  // Authenticate using the first token found.
+  this._authenticateGitHub();
+
+  // Request and save the user ID's and channel ID's for subsequent calls to the Slack API.
+  // This returns a promise, so if you wanted to just send a message to a particular user
+  // you would do:
+  //
+  // this._getSlackMetadata()
+  // .then(function(){
+  //   that.postMessage(this._userIDs['oshehata'], 'Hello ' + this._userData[this._userIDs['oshehata']].real_name);
+  // });
+  //
+  // Which would print 'Hello Omar Shehata' to @oshehata.
+  this._getSlackMetadata();
+};
+
+SlackBot.initializeScheduledJobs = function() {
+  var that = this;
+
+  // Set up the release reminders check every morning at 8 am.
+  schedule.scheduleJob('0 8 * * *', function(){
+    that._sendReleaseReminders();
+  });
+
+  // Set up weekly Slack message every Friday morning.
+  schedule.scheduleJob('0 8 * * FRI', function(){
+    that._sendWeeklyStats();
+  });
+};
+
+SlackBot._getAllIssuesLastWeek = function(repositoryNames) {
+  // Given a list of repository names, return a list of promises that resolve to
+  // a list of all closed issues/PR's last week.
+  var promisesArray = [];
+  for (var i = 0; i < repositoryNames.length; ++i) {
+      var nameSplit = repositoryNames[i].split('/');
+      var owner = nameSplit[0];
+      var repo = nameSplit[1];
+      // Get all closed issues/PRs for this repo for last week.
+      var today = new Date();
+      var lastWeek = new Date(today.getFullYear(), today.getMonth(), today.getDate() - 7);
+      var options = octokit.issues.listForRepo.endpoint.merge({ owner: owner, repo: repo, since: lastWeek.toISOString(), state: 'closed' });
+      promisesArray.push(octokit.paginate(options));
+    }
+
+  return promisesArray;
+};
+
+SlackBot._sendWeeklyStats = function(disableRandomGreeting) {
+    var repositoryNames = Object.keys(this._repositories);
+    var promises = this._getAllIssuesLastWeek(repositoryNames);
+
+    var longestMergeTime = 0;
+    var longestPullRequest;
+    var mergeTimes = [];
+    var that = this;
+
+    return Promise.all(promises)
+    .then(function(values) {
+      for(var i = 0; i < values.length; ++i) {
+        var listOfIssues = values[i];
+        for(var j = 0; j < listOfIssues.length; ++j) {
+          var issue = listOfIssues[j];
+
+          if (defined(issue.pull_request)) {
+            var endTime = new Date(issue.closed_at);
+            var startTime = new Date(issue.created_at);
+            var daysDifference = (endTime - startTime) / (24 * 60 * 60 * 1000);
+            mergeTimes.push(daysDifference);
```

Comment:
I've refactored it to use Promise.each, which now I think is a lot cleaner. I don't think there's a good way to pass those variables from the first block to the next, because Promise.each's `then` will resolve with the original array after they're all done.

I'm open to suggestions though.